src/components/Avatar/Avatar.Image.tsx

```diff
@@ -1,68 +1,263 @@
+      this.props.onLoad && this.props.onLoad()
+    }
+  }
+
+  onError = () => {
+    cache[this.sourceList[this.state.currentIndex]] = false
+    // if the current image has already been destroyed, we are probably no longer mounted
+    // no need to do anything then
+    /* istanbul ignore else */
+    if (!this.image) return false
+
+    // before loading the next image, check to see if it was ever loaded in the past
+    for (
+      var nextIndex = this.state.currentIndex + 1;
+      nextIndex < this.sourceList.length;
+      nextIndex++
+    ) {
+      // get next img
+      let src = this.sourceList[nextIndex]
+
+      // if we have never seen it, its the one we want to try next
+      if (!(src in cache)) {
+        this.setState({ currentIndex: nextIndex })
+        break
+      }
+
+      // if we know it exists, use it!
+      /* istanbul ignore else */
+      if (cache[src] === true) {
+        this.setState({
+          currentIndex: nextIndex,
+          isLoading: false,
+          isLoaded: true,
+        })
+        this.props.onLoad && this.props.onLoad()
+        return true
+      }
+
+      // if we know it doesn't exist, skip it!
+      /* istanbul ignore next */
+      if (cache[src] === false) continue
+    }
+
+    // currentIndex is zero bases, length is 1 based.
+    // if we have no more sources to try, return - we are done
+    if (nextIndex === this.sourceList.length) {
+      this.setState({ isLoading: false })
+      this.props.onError && this.props.onError()
+      return false
+    }
 
-AvatarImage.defaultProps = {
-  animationDuration: 160,
-  animationEasing: 'ease',
-  hasImage: false,
-  image: null,
-  isImageLoaded: false,
-  onError: noop,
-  onLoad: noop,
-  name: null,
-  title: null,
+    // otherwise, try the next img
+    this.loadImg()
+    return false
+  }
+
+  loadImg = () => {
+    this.image = new Image()
+    this.image.src = this.sourceList[this.state.currentIndex]
+
+    if (this.image.decode) {
+      this.image
+        .decode()
+        .then(this.onLoad)
+        .catch(this.onError)
+    } else {
+      this.image.onload = this.onLoad
+      this.image.onerror = this.onError
+    }
+  }
+
+  unloadImg = () => {
+    this.image.onerror = null
+    this.image.onload = null
+
+    // abort any current downloads https://github.com/mbrevda/react-image/pull/223
+    this.image.src = ''
+
+    try {
+      delete this.image.src
+    } catch (e) {
+      // On Safari in Strict mode this will throw an exception,
+      //  - https://github.com/mbrevda/react-image/issues/187
+      // We don't need to do anything about it.
+    }
+    delete this.image
+  }
+
+  componentDidMount() {
+    // kick off process
+    /* istanbul ignore else */
+    if (this.state.isLoading) this.loadImg()
+  }
+
+  componentWillUnmount() {
+    // ensure that we dont leave any lingering listeners
+    /* istanbul ignore else */
+    if (this.image) this.unloadImg()
+  }
+
+  componentWillReceiveProps(nextProps) {
```

Comment:
Just a note, I'm not entirely sure what we are going to do if they actually remove this method from React, `getDerivedStateFromProps` does not behave like this one does 😕

Reply:
good question, I am not sure either. We could use a variable just like the cache one, generate an id for each avatar, and store the image src using that id

we'll see when we upgrade to react 17 😄